README.md
<h1> ESPnow MQTT Home Assistant Gateway </h1>

<h2> Idea:</h2>
<ul>
  <li>Build universal sensor device (ESP32 based) equipped with temperature, humidity and light sensors, that can live long on 1 battery charge</li>
  <li>Clone/deploy it to as many sensors as needed (indoor, outdoor etc.) without changing the software or reconfiguration (beyond what is specific to the device i.e. GPIO connection etc.)
  <li>Build gateway device (that is always powered ON) that transfers data from sensor devices to Home Assistant and automatically creates devices/entities on Home Assistant (no configuration needed) - <b>EDIT:</b> gateway device is now "devices" and every sensor device sends ESPnow message to BROADCAST address. Building multiple gateway devices (and locating them in various places) allows widening the range, without loosing the data when RSSI is very low (extending the range).
</ul>
<br>
<h3>Tasks for the devices:</h3>
<ul>
  <li>Sensor devices: (battery powered) wake up, measure the environment and battery, send to gateway over ESPnow, go to sleep, wake up after specified period and repeat
  <li>Gateway device: (always ON) receive message over ESPnow from sensors, convert to Home Assistant auto discovery code and send to Home Assistant over MQTT/WiFi
  <s><p><b>EDIT:</b>Gateway device, since it is always powered ON, got also motion detector (microwave) so it can be also a part of the home alarm system</s> - REMOVED due to interferences with WiFi
</ul>
<br>
<h3> To satisfy the requirements I've chosen: </h3>
<ul>
  <li>SHT31 temperature and humidity sensor, SDA, around 1.5$ on Aliexpress (breakout)
  <li>TSL2561 light sensor, SDA, around 1.5$ on Aliexpress (breakout)
  <li>MAX17048 LiPo battery sensor, SDA 1-5$ (depending if chip only or breakout)
  <li>ESP32S-WROOM or ESP32-S2 WROOM, around 2$ on Aliexpress (bare metal - <b>do NOT use the development board with additional components! - they cannot go to uA due to the on board elements, i.e. LDO, LED etc.)</b>
  <li>few additional components as per schematics (see below)
</ul>
<br>
<h3>Final schematic (ESP32-S2 based):</h3>
<img width="1088" alt="Screenshot 2022-07-07 at 21 45 12" src="https://user-images.githubusercontent.com/46562447/177868190-8f3dac36-4d2f-49e9-b07d-a71ff26dd909.png">

<br>
<br>
<br>
<h3>Final PCB (ESP32-S2 based) - design:</h3>
<img width="1299" alt="Screenshot 2022-07-07 at 21 27 09" src="https://user-images.githubusercontent.com/46562447/177865693-10016c07-eb08-447c-9519-59ccc6162da4.png">
<br>
<img width="1037" alt="Screenshot 2022-07-07 at 21 36 29" src="https://user-images.githubusercontent.com/46562447/177866919-afb88319-9ee9-4c2e-9bfc-9122b748b7b3.png">

<br>
<h3>Preparing the boards to solder:</h3>
<img width="1221" alt="preparation" src="https://user-images.githubusercontent.com/46562447/180612901-c6468ebf-43ef-4aff-9153-c8fa3c386690.png">



<br>
<h3>Soldered on hot plate:</h3>
<img width="1004" alt="soldered" src="https://user-images.githubusercontent.com/46562447/180612912-b02b499b-4335-4df5-aa38-8d840106d9a7.png">

<br>
<h3>SHT31 - temperature and humidity sensor:</h3>
<img width="721" alt="sht" src="https://user-images.githubusercontent.com/46562447/180615851-c810309d-3509-4d79-b8c0-36f1eb9ef930.png">

<br>
<h3>TSL2561 - light sensor:</h3>
<img width="731" alt="tsl" src="https://user-images.githubusercontent.com/46562447/180615863-af1f1b5c-3b35-44e4-bff1-814092fb4a18.png">

<br>
<h3>Testing all before putting into the box:</h3>
<img width="1158" alt="testing" src="https://user-images.githubusercontent.com/46562447/180615874-d182682f-36ac-4bc9-83c6-fdc08481ce72.png">

<br>
<br>
<h2>Power consumption and management</h2>
<h3> Sensor device is powered with LiPo battery and equipped with TP4056 USB-C charger.</h3>
It can be connected to solar panel or - from time to time if possible/needed - to USB-C charger.
<br>
To minimise the sleep current, the power for all sensors is drawn from one of the ESP32 GPIO, so during the sleep time there is no current leakage - sensors are not powered up.
With the above configuration the sleep current is as following (measured with PPK2):
<ul>
  <li>ESP32-S  WROOM - 4.5uA
  <li>ESP32-S2 WROOM - 22.5uA
</ul>
You would say: "ok, ESP32-S is the winner!" but wait, working time and current are as following:
<ul>
  <li>ESP32-S  WROOM - 600ms, 50mA average
  <li>ESP32-S2 WROOM - 200ms, 38mA average
</ul>

ESP32-S WROOM working time and current:
<br>
<img width="1635" alt="s working time 1-2022-06-17 at 20 33 04" src="https://user-images.githubusercontent.com/46562447/174502504-b3937797-6bf0-40bb-af8e-29bc7e98cdff.png">

ESP32-S2 WROOM working time and current:
<br>
<img width="1667" alt="s2 working time 1-2022-06-18 at 12 51 43" src="https://user-images.githubusercontent.com/46562447/174502510-bb3200b7-647a-4269-9ca5-2978d3c97793.png">


And that is what really matters with the battery life time calculation, because 3 times shorter working time makes the difference.<br>
<h4>Rough and rounded up calculation (1000mAh battery, reserve capacity=20%, working time as above, sleep time=180s) shows as following:</h4>
<ul>
  <li>ESP32-S  WROOM: 4 700 hours with average 170uA = 195 days, 6.5 months
  <li>ESP32-S2 WROOM: 12 370 hours with average 65uA = 515 days, 17 months
</ul>

<br>
<b>ESP32-S WROOM battery life time calculation:</b>
<br>

<img width="454" alt="Screenshot 2022-06-19 at 22 20 17" src="https://user-images.githubusercontent.com/46562447/174500944-ad46fd8c-fa2e-4983-8f97-e45409a844da.png">


<br>
<b>ESP32-S2 WROOM battery life time calculation:</b>
<br>

<img width="435" alt="Screenshot 2022-06-19 at 22 20 59" src="https://user-images.githubusercontent.com/46562447/174500948-ce6a89dc-d225-4094-a5b6-923444074d7f.png">
<br>
So apparently the winner is <b>ESP32-S2 WROOM</b> with almost triple battery life.
<p>
<b>EDIT:</b> Final product as per design above consumes while sleeping 33uA - I have included LDO to make sure I have stable 3.3V - unfortunately it takes 8uA - still, as described above, the clue is in the very short working time rather than few uA here or there during the sleep - we are still talking above 9 200 hours on 1 battery charge - more than a year (383 days).
<p>
<br>
  <img width="470" alt="Screenshot 2022-07-23 at 18 24 08" src="https://user-images.githubusercontent.com/46562447/180616082-d279e82b-7feb-4438-9cb7-816fcb115e12.png">

<br>
<br>
<h3>Measuring the working time</h3>
To measure working time you shall NOT rely only on millis or micros. I.e. ESP32S reports millis as first line in setup() as xx while ESP32-S2 shows 280ms.
And actually both are wrong as you can see below on the screens:
<br>
<br>
<b>ESP32S times measured:</b>
<br>
<img width="1728" alt="esp32s" src="https://user-images.githubusercontent.com/46562447/175294471-01d2a13f-5269-493d-ad02-d072cb563ad3.png">

<br>
<br>
<br>
<b>ESP32S-2 times measured. And on the left side the correct ontime reported by the sketch with error correction applied:</b>
<br>
<br>
<img width="1707" alt="s2" src="https://user-images.githubusercontent.com/46562447/175366621-a6fbe7bb-2338-48c7-8939-bfce6faa7541.png">

<br>
<br>
I used PPK2 and estimated the time the ESP32 works measuring the power consumption.
<br>
<br>
<b>Conclusions:</b>
<ul>
  <li>don't rely on millis() only
  <li>ESP32-S2 starts much faster (35ms) than ESP32S (310ms)
  <li>ESP32-S2 total working time is much shorter so the total power consumption with S2 is much lower accordingly - battery life is extended
</ul>
<br>
<br>
<b>ESP32S and ESP32-S2 ontime after 1.5 days with 180s sleep - almost 2.5-3x more "ontime" for ESP32S:</b>
<br>
<img width="1556" alt="Screenshot 2022-07-01 at 09 51 50" src="https://user-images.githubusercontent.com/46562447/176860918-338ccdcc-a744-4832-8502-c4c54a152008.png">
<br>
<br>
<img width="809" alt="Screenshot 2022-07-01 at 09 54 44" src="https://user-images.githubusercontent.com/46562447/176861422-8171adab-98db-4501-8c49-6f7d3a2e9f59.png">

<br>
<br>


<h3>Charging details</h3>
Sensor device also provides information about charging status:
<ul>
  <li>NC - not connected
  <li>ON - charging
  <li>FULL - charged
</ul>  
To achieve this, you need to connect the pins from TP4056 that control charging/power LEDs - see below - to GPIO of ESP32 - see the schematics above. <br>
<br>
<img width="643" alt="Screenshot 2022-06-19 at 21 39 22" src="https://user-images.githubusercontent.com/46562447/174499779-ea0c5474-8ac7-4ac6-acd1-df2f67e84a80.png">

<img width="626" alt="Screenshot 2022-06-19 at 21 40 27" src="https://user-images.githubusercontent.com/46562447/174499788-fd8f5c83-6684-4a11-8a63-4529930a9508.png">

<h2>Firmware update - OTA</h2>
<h3>Sensor device</h3>
To perform firmware update there are 3 possibilities - 2 are with web server in use, where you store the binary file (sender.ino.esp32.bin)
<ul>
  <li>double reset click - built in functions recognise double reset and if so done, performs firmware update - of course you need to visit the sensor to double click it ;-)
  <li>routine check for new firmware availability on the server: every 24h (configurable) sensor device connects to server and if new file found, performs update
  <li>3rd: wire TX/RX to the FTDI programmer
</ul>
Sensor device during firmware update:<br>
<br>
<img width="540" alt="Screenshot 2022-06-22 at 18 55 39" src="https://user-images.githubusercontent.com/46562447/175105073-ad607c41-860e-4edd-8235-77bd8709eb10.png">
<br>
<h3>Gateway device</h3>
To perform firmware update you simply click the button "Update" on Home Assistant (in the device section of gateway) - gateway will connect to the server where the binary is stored and if file is found, it will perform firmware update and restart gateway.
I am using Apache minimal add on Home Assistant - since all sensors are in the same network where Home Assistant is, <b>there is no need for internet access for sensors (and gateway).</b>
<br>

<h2>Configuration</h2>
All sensors used in the sensor device (SHT31, TSL2561, MAX17048, checking charging status) are optional and can be disabled in configuration file.
On top of that, light sensor (TSL2561) can be replaced with phototransistor (i.e. TEPT4400) - also configurable.
In the simplest (and useless) configuration, sensor sends only... its name to Home Assistant.

Sleep time (configurable) is initially set to 180s (3 minutes).

Other important configurable settings (some mandatory, some optional):
<ul>
  <li>WiFi channel - it must be the same for ESPnow communication and gateway to AP (router) communication - gateway and sensor devices
  <li>SSID - gateway and sensor devices
  <li>password - gateway and sensor devices
  <li>MQTT server IP - gateway device
  <li>MQTT username - gateway device
  <li>MQTT password - gateway device
</ul>
<h2>Software</h2>
<h3>Sensor device - sender</h3>
Tasks are as described above so the code is in 1 file only: sender.ino <br>
<br>
Next is the file with credentials: passwords.h (ssid, password, webserver where your firmware is stored) - this file is only used for OTA.<br>
<br>
Finally there is a configuration file where you must specify details for each sensor device (such as sensors used, GPIO etc.): devices_config.h<br>
<br>
The sequence is:
<ul>
  <li>configure the device in devices_config.h
  <li>uncomment the DEVICE_ID you want to compile the sketch for in main sketch: sender.ino
  <li>compile/upload to the device
  <li>repeat the above 3 points for the next sensor device
</ul>

<br>
Sender in action (test device):<br>
<br>
<img width="484" alt="Screenshot 2022-06-22 at 18 56 08" src="https://user-images.githubusercontent.com/46562447/175104814-a21c53e7-a631-441d-b5cc-1bffd557b9a9.png">
<br>

<h3>Gateway device - receiver</h3>
Gateway tasks are more complex (as described above) so the code is split into multiple files - per function<br>
Entire configuration is in config.h file<br>
<br>
Next is the file with credentials: passwords.h (ssid, password, mqtt ip and credentials, webserver where your firmware is stored)
<br>

<br>
Receiver in action (real device):<br>
<br>
<img width="529" alt="Screenshot 2022-06-22 at 18 41 34" src="https://user-images.githubusercontent.com/46562447/175102381-c1ade15e-66b6-44f4-95fa-aba3ba5a88b8.png">

<br>

<h3>Libraries needed (non standard):</h3>
<ul>
  <li>ESP_DoubleResetDetector
  <li>HTTPClient
  <li>Update
  <li>Adafruit_SHT31
  <li>SparkFunTSL2561
  <li>SparkFun_MAX1704x_Fuel_Gauge_Arduino_Library
  <li>PubSubClient
  <li>ArduinoJson
</ul>

<h2>Home Assistant</h2>
<h3>Information from sensor device on Home Assistant:</h3>
<ul>
  <li>device name
  <li>temperature [C]
  <li>humidity [%]
  <li>light [lux]
  <li>battery in [%] and [Volt]
</ul>

Additionally few diagnostic information:
<ul>
  <li>wifi signal/rssi [dBm]
  <li>firmware version
  <li>boot number
  <li>[cumulative] ontime (in seconds) for further postprocessing i.e. in Grafana - it resests to 0 when charging starts
  <li>[cumulative] pretty ontime, i.e.: "1d 17h 23m 12s" (human readable) - it resests to 0 when charging starts
</ul>
<br>
<b>Ontime is calculated</b> based on millis (time just before hibernation - start time) including empirically established difference between real startup time and measured with PPK2 (apparently both: ESP32 and ESP32-S2 are "cheating" with millis, however S2 shows hundreds of millis just after start, while ESP32 shows very low value after start - both are innaccurate and mainly depend on the size of the binary file, that has to be loaded into ESP32 memory during startup by bootloader).
Ontime is also reduced by turning off the bootloader logo during wake up from sleep.
<h4>Gateway device (with its entities) as well as all sensor devices (with their entities) are automatically configured in Home Assistant using MQTT discovery</h4>
<br>
<br>
<b>Gateway device as discovered by Home Assistant:<br>
<br>

<img width="532" alt="Screenshot 2022-07-04 at 09 11 38" src="https://user-images.githubusercontent.com/46562447/177111787-2e294347-cb40-4041-80e7-a02f0d6d7af2.png">

<br><br>
<b>Gateway device on lovelace dashboard:<br>
<br>

<img width="390" alt="Screenshot 2022-07-04 at 09 10 54" src="https://user-images.githubusercontent.com/46562447/177111873-b1d236dd-abfc-4cbf-91e7-c3a316369cee.png">

<br><br>
<b>Sensor device as discovered by Home Assistant:<br>
<br>

<img width="531" alt="Screenshot 2022-07-04 at 09 11 52" src="https://user-images.githubusercontent.com/46562447/177111944-b11afa4f-93ac-4319-966b-8fcfd8772ad1.png">

<br><br><br>
<b>Sensor Device on lovelace dashboard:<br>
<br>

<img width="390" alt="Screenshot 2022-07-04 at 09 11 21" src="https://user-images.githubusercontent.com/46562447/177112009-923aa439-4b00-4d8e-9aac-3a878f2b3087.png">

<br>
<h3>MQTT structure</h3>
All information from sensor device is sent to Home Assistant MQTT broker in one topic: sensor_hostname/sensor/state<br>
<br>
<img width="593" alt="Screenshot 2022-06-20 at 00 17 05" src="https://user-images.githubusercontent.com/46562447/174503988-5944ceb1-49b5-40a9-a96f-f9aeae532dee.png">

<br>And message arrives in JSON format:<br>
<br>
<img width="628" alt="Screenshot 2022-06-22 at 19 26 51" src="https://user-images.githubusercontent.com/46562447/175109966-148e09bf-1cd4-48b1-bd95-0f66ca9b1206.png">


<br>
<h2>TODO list</h2>
<ul>
  <li>Sensor device: OTA - how to do it ON DEMAND? No idea yet. If device is to listen to gateway, it will consume battery... Anybody knows?
</ul>


sender/changelog.txt
2022-07-30:
  1.16.1  - new devices: 100. 101, 102, 104
          - PERIODIC_FW_CHECK_HRS  14400 -> 30 days between updates

2022-07-26:
  1.16.0  - change of the data structure
          - lux_with_tept(long int* lux_int); - not char

2022-07-24:
  1.15.2  - MINIMUM_VOLTS=3.3V for new devices

2022-07-24:
  1.15.1a - lipo.reset() on reset button or power reconnect
          - boot_reason in loop() as sometimes Serial is not fast enough to print in setup()

2022-07-23:
  1.15.0 - "motion" "env" ok
         - corrections

2022-07-06:
  1.15.b1 - added:
            char sender_type[10]; // "motion" "env"
            char motion[2];       // "0" "1"  [2]
            for motion devices to reuse gateway with the same message structure

2022-07-06:
  1.14.2  - default PERIODIC_FW_CHECK_HRS is now 168 - 7 days
  1.14.1  - turn off ENABLE_3V_GPIO when gather_data() fails (otherwise sleep current is 300uA)

2022-07-05:
  1.14.0  - rename LEDs, blink RED on error, MINIMUM_VOLTS=3.5V

2022-07-04:
  1.14.b1 - battery change rate, redo myData structure, fake receiver, lipo.quickStart abandoned

2022-07-02:
  1.13.0 - define MAC_BROADCAST for sending to all (receivers)

2022-07-02:
  1.12.0 - send to either of 2 receivers depending on bootCount (even/odd)

2022-06-28:
  1.11.1 - boot_reason added to reset bootCount on button reset
  1.11.0 - major rework on charging_state()
         - cosmetics

2022-06-25:
  1.10.5 - blink FW_UPGRADE_LED_GPIO on FAILED
  1.10.4 - PERIODIC_FW_CHECK changed to PERIODIC_FW_CHECK_HRS

2022-06-24:
  1.10.3 - print MAC address while connecting to AP
  1.10.2 - #define GND_GPIO_FOR_LED      13    // if not equipped comment out - GND for ACTIVITY_LED_GPIO
         - testing esp_wifi_set_protocol(WIFI_IF_STA, WIFI_PROTOCOL_11B|WIFI_PROTOCOL_11G|WIFI_PROTOCOL_11N|WIFI_PROTOCOL_LR);
           to extend the range of radio (no conclusion yet)
         - print MAC address while connecting to AP

2022-06-23:
  1.10.1 - ESP32_IS_CHEATING measured for ESP32-S2 to be -288ms (head, tail, last commands) - added and verified

2022-06-23:
  1.10.0 - PPK2_GPIO and code added to measure current and time
         - ESP32_IS_CHEATING measured for ESP32S to be 325ms (head, tail, last commands) - added and verified

2022-06-22:
  1.9.2 - version description in changelog.txt

2022-06-22:
  1.9.1 - cosmetics

2022-06-22:
  1.9.0 - ontime saved
        - define ESP32_IS_CHEATING

2022-06-20:
  1.8.2 - cleaning even more: printf instead of print(ln) where possible
        - reset bootCount on DRD_Detected

2022-06-19:
  1.8.1 - cleaning code for github ;-)

2022-06-18:
  1.8.0 - bootCount in file and in HA, periodic FW upgrade/check every PERIODIC_FW_CHECK (number of restarts)

2022-06-18:
  1.7.1 - "FW UPG..." instead of VERSION during upgrade

2022-06-16:
  1.7.0 - major rework of wire.h etc. - should be ok now

2022-06-16:
  1.6.1 - choose device to be compiled in main sketch

  1.6.0 -  TSL2561 moved to SparkFunTSL2561, all in gather_data - nothing in setup
        -  see if battery is still up/down every 30 minutes

2022-06-15:
  1.5.1 - move boards definitions to "devices_config.h"
        - supress boot messages from ESP

2022-06-14:
  1.5.0 - Major redefinitions, choose sensors the device is equipped with

2022-06-14:
  1.4.0 - MINIMUM_VOLTS added to avoid start wifi if too low battery
        - WAIT_FOR_WIFI=5s in seconds for upgrade

2022-06-12:
  1.3.1 - added USE_CUSTOM_I2C_GPIO
        - moved SLEEP_TIME to each sensor config to have different sleep times

2022-06-12:
  1.3.0 - added DEVICE_NAME

2022-06-12:
  1.2.0 - Firmware Upgrade on double push button (fast after boot (led blinks))
        - FW_UPGRADE_LED_GPIO (flashing during firmware upgrade) - red LED  flashes during upgrade, SOS on failure
        - if no FW_UPGRADE_LED_GPIO then ACTIVITY_LED_GPIO will flash (if defined)

2022-06-11:
  1.1.0 - added lux from TEPT/ADC, added CHARGING_GPIO


sender/devices_config.h
#pragma once
/*
last updated to the new version: esp32100

device configuration template:

#define HOSTNAME              "esp32028"   // obligatory, max 9 characters
#define DEVICE_NAME           "Garage"     // obligatory, max 15 characters
#define BOARD_TYPE            1     // obligatory, 1=ESP32/WROOM,WROVER 2=ESP32S2 3=ESP32S3 4=ESP32C3
#define MINIMUM_VOLTS         3.3   // minimum voltage to start wifi, if not defined in device config then it is done in sender.ino
#define ENABLE_3V_GPIO        21    // if not equipped comment out - power for sensors from Vcc
#define ACT_BLUE_LED_GPIO     25    // if not equipped comment out - it blinks when sensor is ON/ if not defined ERROR_RED_LED_GPIO then it blinks when upgrade firmware is ongoing or on error
#define ERROR_RED_LED_GPIO    26    // if not equipped comment out - blinks when upgrade firmware is ongoing and or error
#define POWER_GPIO            38    // pin 6, green LED on charger TP4056, HIGH on charging, LOW on full
#define CHARGING_GPIO         39    // pin 7, red   LED on charger TP4056, HIGH on full, LOW on charging
#define SLEEP_TIME            180   // seconds
#define USE_MAX17048          1     // fuel gauge - optional, 0 if not in use
#define USE_SHT31             1     // temp/hum - optional, 0 if not in use
#define USE_TSL2561           0     // lux SDA - optional, 0 if not in use
#define USE_TEPT4400          1     // lux ADC - optional, 0 if not in use
#define LUX_ADC_GPIO          36    // relevant only if USE_TEPT4400=1
#define LUX_MAX_RAW_READING   2828  // relevant only if USE_TEPT4400=1
#define USE_CUSTOM_I2C_GPIO   1     // instead of standard ones: 21,22 for ESP32S and 8,9 for ESP32S2), 0 if not in use
#define SDA_GPIO              18    // relevant only if USE_CUSTOM_I2C_GPIO=1
#define SCL_GPIO              19    // relevant only if USE_CUSTOM_I2C_GPIO=1
#define GND_GPIO_FOR_LED      13    // if not equipped comment out - GND for ACT_BLUE_LED_GPIO
#define PERIODIC_FW_CHECK_HRS 24    // check FW update every hours, if not defined in device config then it is done in sender.ino

test devices: esp32080 - 099
*/


// *************** "esp32028" - S,  production - Garage ************************
#if   DEVICE_ID == 28
  #define HOSTNAME              "esp32028"
  #define DEVICE_NAME           "Garage"
  #define BOARD_TYPE            1
  #define ENABLE_3V_GPIO        21
  #define ACT_BLUE_LED_GPIO     25
  #define ERROR_RED_LED_GPIO    26
  #define CHARGING_GPIO         35
  #define POWER_GPIO            32
  #define SLEEP_TIME            180
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           0
  #define USE_TEPT4400          1
  #define LUX_ADC_GPIO          36
  #define LUX_MAX_RAW_READING   2828
  #define USE_CUSTOM_I2C_GPIO   1
  #define SDA_GPIO              18
  #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32028-Garage"

// *************** "esp32100" - S2, production - Papa new ************************
#elif DEVICE_ID == 100
  #define HOSTNAME              "esp32100"
  #define DEVICE_NAME           "Papa"
  #define BOARD_TYPE            2
  #define ENABLE_3V_GPIO        3
  #define ACT_BLUE_LED_GPIO     6
  #define ERROR_RED_LED_GPIO    5
  #define CHARGING_GPIO         39
  #define POWER_GPIO            38
  #define SLEEP_TIME            180
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           1
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          13
  #define LUX_MAX_RAW_READING   2900
  // #define USE_CUSTOM_I2C_GPIO   0
  // #define SDA_GPIO              18
  // #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32100-Papa-NEW device"

// *************** "esp32101" - S2, production - Dining new ************************
#elif DEVICE_ID == 101
  #define HOSTNAME              "esp32101"
  #define DEVICE_NAME           "Dining"
  #define BOARD_TYPE            2
  #define ENABLE_3V_GPIO        3
  #define ACT_BLUE_LED_GPIO     6
  #define ERROR_RED_LED_GPIO    5
  #define CHARGING_GPIO         39
  #define POWER_GPIO            38
  #define SLEEP_TIME            180
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           1
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          13
  #define LUX_MAX_RAW_READING   2900
  // #define USE_CUSTOM_I2C_GPIO   0
  // #define SDA_GPIO              18
  // #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32101-Dining-NEW device"

// *************** "esp32102" - S2, production - Toilet up new ************************
#elif DEVICE_ID == 102
  #define HOSTNAME              "esp32102"
  #define DEVICE_NAME           "Toilet"
  #define BOARD_TYPE            2
  #define ENABLE_3V_GPIO        3
  #define ACT_BLUE_LED_GPIO     6
  #define ERROR_RED_LED_GPIO    5
  #define CHARGING_GPIO         39
  #define POWER_GPIO            38
  #define SLEEP_TIME            180
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           1
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          13
  #define LUX_MAX_RAW_READING   2900
  // #define USE_CUSTOM_I2C_GPIO   0
  // #define SDA_GPIO              18
  // #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32102-Toilet up-NEW device"


// *************** "esp32104" - S2, production - Milena new ************************
#elif DEVICE_ID == 104
  #define HOSTNAME              "esp32104"
  #define DEVICE_NAME           "Milena"
  #define BOARD_TYPE            2
  #define ENABLE_3V_GPIO        3
  #define ACT_BLUE_LED_GPIO     6
  #define ERROR_RED_LED_GPIO    5
  #define CHARGING_GPIO         39
  #define POWER_GPIO            38
  #define SLEEP_TIME            180
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           1
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          13
  #define LUX_MAX_RAW_READING   2900
  // #define USE_CUSTOM_I2C_GPIO   0
  // #define SDA_GPIO              18
  // #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32102-Milena-NEW device"

// *************** "esp32105" - S2, production - Garden ************************
#elif DEVICE_ID == 105
  #define HOSTNAME              "esp32105"
  #define DEVICE_NAME           "Garden"
  #define BOARD_TYPE            2
  #define ENABLE_3V_GPIO        3
  #define ACT_BLUE_LED_GPIO     6
  #define ERROR_RED_LED_GPIO    5
  #define CHARGING_GPIO         38
  #define POWER_GPIO            39
  #define SLEEP_TIME            180
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           1
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          36
  #define LUX_MAX_RAW_READING   2900
  #define USE_CUSTOM_I2C_GPIO   0
  #define SDA_GPIO              18
  #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32105-Garden"

// =======================================================================================================

// *************** "esp32086" - S2, new_device template -  ****************************
#elif DEVICE_ID == 86
  #define HOSTNAME              "esp32086"
  #define DEVICE_NAME           "new_dev"
  #define BOARD_TYPE            2
  #define MINIMUM_VOLTS         3.3
  #define ENABLE_3V_GPIO        3
  #define ACT_BLUE_LED_GPIO     6
  #define ERROR_RED_LED_GPIO    5
  #define CHARGING_GPIO         39
  #define POWER_GPIO            38
  #define SLEEP_TIME            5
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           1
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          13
  #define LUX_MAX_RAW_READING   2900
  // #define USE_CUSTOM_I2C_GPIO   0
  // #define SDA_GPIO              18
  // #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32086-new_device S2 template"
//

// *************** "esp32092" - S3,  test - S3 ***********************************

#elif DEVICE_ID == 92
  #define HOSTNAME              "esp32092"
  #define DEVICE_NAME           "testS3"
  #define BOARD_TYPE            3
  // #define ENABLE_3V_GPIO        32
  #define ACT_BLUE_LED_GPIO     7
  #define ERROR_RED_LED_GPIO    5
  // #define CHARGING_GPIO         38
  // #define POWER_GPIO            39
  #define SLEEP_TIME            5
  #define USE_MAX17048          0
  #define USE_SHT31             0
  #define USE_TSL2561           0
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          36
  #define LUX_MAX_RAW_READING   2900
  #define USE_CUSTOM_I2C_GPIO   0
  #define SDA_GPIO              18
  #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32092-testS3"

// *************** "esp32093" - S - old ************************
#elif DEVICE_ID == 93
  #define HOSTNAME              "esp32093"
  #define DEVICE_NAME           "testS"
  #define BOARD_TYPE            1
  #define ENABLE_3V_GPIO        19
  #define ACT_BLUE_LED_GPIO     25
  #define ERROR_RED_LED_GPIO    26
  #define CHARGING_GPIO         16
  #define POWER_GPIO            17
  #define SLEEP_TIME            5
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           1
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          4
  #define LUX_MAX_RAW_READING   7500
  #define USE_CUSTOM_I2C_GPIO   0
  #define SDA_GPIO              18
  #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32093-S-old "


// *************** "esp32094" - S2, old  ************************ - based on garden S2 105
#elif DEVICE_ID == 94
  #define HOSTNAME              "esp32094"
  #define DEVICE_NAME           "testS2"
  #define BOARD_TYPE            2
  #define ENABLE_3V_GPIO        3
  #define ACT_BLUE_LED_GPIO     6
  #define ERROR_RED_LED_GPIO    5
  #define CHARGING_GPIO         38
  #define POWER_GPIO            39
  #define SLEEP_TIME            180
  #define USE_MAX17048          1
  #define USE_SHT31             1
  #define USE_TSL2561           1
  #define USE_TEPT4400          0
  #define LUX_ADC_GPIO          36
  #define LUX_MAX_RAW_READING   2900
  #define USE_CUSTOM_I2C_GPIO   0
  #define SDA_GPIO              18
  #define SCL_GPIO              19
  // #define GND_GPIO_FOR_LED      13
  // #define PERIODIC_FW_CHECK_HRS
  #pragma message "compilation for: esp32094-S2-old "
// ---------------------------------------------------------------------------------------------------

#else
  #error "Wrong DEVICE_ID chosen"
#endif


// assigning MODEL and checking if proper board is selected
#ifndef BOARD_TYPE
  #error BOARD_TYPE not defined
#else
  #if (BOARD_TYPE == 1) and (!defined(CONFIG_IDF_TARGET_ESP32))
    #error wrong board selected in Arduino - choose ESP32DEV
  #endif
  #if (BOARD_TYPE == 2) and (!defined(CONFIG_IDF_TARGET_ESP32S2))
    #error wrong board selected in Arduino - choose S2
  #endif
  #if (BOARD_TYPE == 3) and (!defined(CONFIG_IDF_TARGET_ESP32S3))
    #error wrong board selected in Arduino - choose S3
  #endif
  #if (BOARD_TYPE == 4) and (!defined(CONFIG_IDF_TARGET_ESP32C3))
    #error wrong board selected in Arduino - choose C3
  #endif

  #if (BOARD_TYPE == 1)
    #define MODEL "ESP32"
    #define ESP32_IS_CHEATING (325) // +310 head, +12 tail, +3ms delay between saved and save_ontime() function start
    #ifdef PRINT_COMPILER_MESSAGES
      #pragma message "chosen BOARD_TYPE = ESP32"
    #endif
  #elif (BOARD_TYPE == 2)
    #define MODEL "ESP32S2"
    #define ESP32_IS_CHEATING (-228)  // -244 head, +12 tail, +4ms delay between saved and save_ontime() function start
    #ifdef PRINT_COMPILER_MESSAGES
      #pragma message "chosen BOARD_TYPE = ESP32S2"
    #endif
  #elif (BOARD_TYPE == 3)
    #define MODEL "ESP32S3"
    #define ESP32_IS_CHEATING 0       // not tested yet
    #ifdef PRINT_COMPILER_MESSAGES
      #pragma message "chosen BOARD_TYPE = ESP32S3"
    #endif
  #elif (BOARD_TYPE == 4)
    #define MODEL "ESP32C3"
    #define ESP32_IS_CHEATING 0       // not tested yet
    #ifdef PRINT_COMPILER_MESSAGES
      #pragma message "chosen BOARD_TYPE = ESP32C3"
    #endif
  #else
    #error BOARD_TYPE not defined
  #endif
#endif
// assigning MODEL and checking if proper board is selected END
